sys/net/link_layer/l2util/l2util.c

```diff
@@ -0,0 +1,197 @@
+        case NETDEV_TYPE_BLE:
+            if (addr_len == sizeof(eui64_t)) {
+                memcpy(iid, addr, sizeof(eui64_t));
+                iid->uint8[0] ^= 0x02;
+                return sizeof(eui64_t);
+            }
+            else {
+                return -EINVAL;
+            }
+#endif  /* MODULE_NORDIC_SOFTDEVICE_BLE */
+#if defined(MODULE_CC110X) || defined(MODULE_NRFMIN)
+        case NETDEV_TYPE_CC110X:
+        case NETDEV_TYPE_NRFMIN:
+            if (addr_len <= 3) {
+                _create_iid_from_short(addr, addr_len, iid);
+                return sizeof(eui64_t);
+            }
+            else {
+                return -EINVAL;
+            }
+#endif  /* defined(MODULE_CC110X) || defined(MODULE_NRFMIN) */
+        default:
+            (void)addr;
+            (void)addr_len;
+            (void)iid;
+#ifdef DEVELHELP
+            LOG_ERROR("gnrc_netif: can't convert hardware address to IID "
+                      "for device type %d\n", dev_type);
+#endif  /* DEVELHELP */
+            assert(false);
+            break;
+    }
+    return -ENOTSUP;
+}
+
+int l2util_ipv6_iid_to_addr(int dev_type, const eui64_t *iid, uint8_t *addr)
+{
+    switch (dev_type) {
+#if defined(MODULE_NETDEV_ETH) || defined(MODULE_ESP_NOW)
+        case NETDEV_TYPE_ETHERNET:
+        case NETDEV_TYPE_ESP_NOW:
+            eui48_from_ipv6_iid((eui48_t *)addr, iid);
+            return sizeof(eui48_t);
+#endif  /* defined(MODULE_NETDEV_ETH) || defined(MODULE_ESP_NOW) */
+#if defined(MODULE_NETDEV_IEEE802154) || defined(MODULE_XBEE)
+        case NETDEV_TYPE_IEEE802154:
+            /* assume address was based on EUI-64
+             * (see https://tools.ietf.org/html/rfc6775#section-5.2) */
+            memcpy(addr, iid, sizeof(eui64_t));
+            addr[0] ^= 0x02;
+            return sizeof(eui64_t);
+#endif  /* defined(MODULE_NETDEV_IEEE802154) || defined(MODULE_XBEE) */
+#ifdef MODULE_NRFMIN
+        case NETDEV_TYPE_NRFMIN:
+            addr[0] = iid->uint8[6];
+            addr[1] = iid->uint8[7];
+            return sizeof(uint16_t);
+#endif  /* MODULE_NETDEV_IEEE802154 */
+#ifdef MODULE_NORDIC_SOFTDEVICE_BLE
+        case NETDEV_TYPE_BLE:
+            memcpy(addr, iid, sizeof(eui64_t));
+            addr[0] ^= 0x02;
+            return sizeof(eui64_t);
+#endif  /* MODULE_NORDIC_SOFTDEVICE_BLE */
+#ifdef MODULE_CC110X
+        case NETDEV_TYPE_CC110X:
+            addr[0] = iid->uint8[7];
+            return sizeof(uint8_t);
+#endif  /* MODULE_CC110X */
+        default:
+            (void)iid;
+            (void)addr;
+#ifdef DEVELHELP
+            LOG_ERROR("gnrc_netif: can't convert IID to hardware address "
+                      "for device type %d\n", dev_type);
+#endif  /* DEVELHELP */
+            assert(false);
+            break;
+    }
+    return -ENOTSUP;
+}
+
+int l2util_ndp_addr_len_from_l2ao(int dev_type,
+                                  const ndp_opt_t *opt)
+{
+    switch (dev_type) {
+#ifdef MODULE_CC110X
```

Comment:
why isn't that handled by/in the driver to avoid all these `#ifdef` here?

Reply:
Because the NDP link-layer address option has nothing to do with a driver (it is link-layer specific stuff for required for the network layer), it would require *every* driver to implement this + there are similar functions that are not device specific (but very much specific to the link-layer the device implements; e.g. EUI-64/IPv6 IID conversion which can apply to both the source address or the destination address of a packet).

Reply:
(edited for more clarity)